### PR TITLE
S3 | EncodeURIComponent / DecodeURIComponent xattr

### PR DESCRIPTION
**Pull request by romayalon**
### Explain the changes
1. Revert the `__dot__` convertion introduced on - https://github.com/noobaa/noobaa-core/pull/7408.
2. PUT Object / Create MPU - Call decodeURIComponent(xattr_val) before processing the request.
3. Head Objcet - If not encoded - Call encodeURIComponent(xattr_val). We shouldn't encode twice an already encoded xattr values (for example - In NSFS flow, in a case where the xattr is set directly on the file) 

### Issues: Fixed #xxx / Gap #xxx
1. Fixed #https://github.com/noobaa/noobaa-core/issues/7404
2. Fixed #https://github.com/noobaa/noobaa-core/issues/7896

### Testing Instructions:
1. `sudo NC_CORETEST=true node ./node_modules/mocha/bin/mocha noobaa-core/src/test/unit_tests/test_bucketspace.js`

- [ ] Doc added/updated
- [x] Tests added

### CODE REVIEW

**Review by tangledbytes — APPROVED** [review]

File: src/endpoint/s3/s3_utils.js [hunk 1]
```diff
@@ -616,6 +613,16 @@ function parse_decimal_int(str) {
     return parsed;
 }
 
+/**
+ * isEncoded consumes returns true if string uri is URIEncoded
+ * @param {string} uri
+ * @returns {boolean}
+ */
+function isEncoded(uri) {
+    uri = uri || '';
```
Comment (guymguym) [inline src/endpoint/s3/s3_utils.js:622]: Use default in function args, renamed, and added also encode_uri_unless_already_encoded() which deserves its own function

```suggestion
function encode_uri_unless_already_encoded(uri = '') {
    return is_uri_already_encoded(uri) ? uri : encodeURIComponent(uri);
}

function is_uri_already_encoded(uri = '') {
```

File: src/endpoint/s3/s3_utils.js [hunk 2]
```diff
@@ -616,6 +613,16 @@ function parse_decimal_int(str) {
...
+    return uri !== decodeURIComponent(uri);
```
Comment (guymguym) [inline src/endpoint/s3/s3_utils.js:623]: nice!

File: src/endpoint/s3/s3_utils.js [hunk 3]
```diff
@@ -85,22 +81,23 @@ function set_response_xattr(res, xattr) {
         keys.sort();
     }
     let returned_keys = 0;
-    for (let key of keys) {
-        // replace characters not allowed by RFC 7230 Section 3.2 with custom strings
-        key = key.replaceAll('.', HTTP_NB_DOT_CHAR);
+    for (const key of keys) {
+        // when xattr is set directly on the object (NSFS for example) and it's already encoded
+        // we should not encode it again 
+        const val = isEncoded(xattr[key]) ? xattr[key] : encodeURIComponent(xattr[key]);
```
Comment (guymguym) [inline src/endpoint/s3/s3_utils.js:87]:
```suggestion
        const val = encode_uri_unless_already_encoded(xattr[key]);
```

**Review by guymguym — APPROVED** [review]
looks good! just style comments

File: src/endpoint/s3/s3_utils.js [hunk 4]
```diff
@@ -57,18 +57,14 @@ function format_s3_xml_date(input) {
...
     _.each(req.headers, (val, hdr) => {
         if (!hdr.startsWith(X_AMZ_META)) return;
-        let key = hdr.slice(X_AMZ_META.length);
+        const key = hdr.slice(X_AMZ_META.length);
         if (!key) return;
-
-        // we encode unacceptable characters in the header name, reverting that here
-        key = key.replaceAll(HTTP_NB_DOT_CHAR, '.');
-        xattr[key] = val;
+        xattr[key] = decodeURIComponent(val);
```
Comment (romayalon) [inline src/endpoint/s3/s3_utils.js:67]: @guymguym WDYT about decoding on PUT?
it's not necessary...

**Pull request merged**